Zend/zend_weakrefs.c

```diff
@@ -0,0 +1,235 @@
+   +----------------------------------------------------------------------+
+   | Copyright (c) The PHP Group                                          |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 2.00 of the Zend license,     |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.zend.com/license/2_00.txt.                                |
+   | If you did not receive a copy of the Zend license and are unable to  |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@zend.com so we can mail you a copy immediately.              |
+   +----------------------------------------------------------------------+
+   | Authors: krakjoe@php.net                                             |
+   +----------------------------------------------------------------------+
+*/
+
+#include "zend.h"
+#include "zend_interfaces.h"
+#include "zend_objects_API.h"
+#include "zend_weakrefs.h"
+
+typedef struct _zend_weakref {
+	zval referent;
+	zend_array *root;
+	zend_object std;
+} zend_weakref;
+
+zend_class_entry *zend_ce_weakref;
+zend_object_handlers zend_weakref_handlers;
+
+#define zend_weakref_from(o) ((zend_weakref*)(((char*) o) - XtOffsetOf(zend_weakref, std)))
+#define zend_weakref_fetch(z) zend_weakref_from(Z_OBJ_P(z))
+
+static inline zend_array* zend_weakrefs_fetch(zend_object *ro, zend_bool create);
+
+static void zend_weakrefs_dtor(zval *zv) {
+	zend_hash_destroy(Z_PTR_P(zv));
+	FREE_HASHTABLE(Z_PTR_P(zv));
+}
+
+void zend_weakrefs_init() {
+	zend_hash_init(&EG(weakrefs), 8, NULL, zend_weakrefs_dtor, 0);
+}
+
+static void zend_weakref_unref(zval *zv) {
+	zend_weakref *wr = (zend_weakref*) Z_PTR_P(zv);
+
+	ZVAL_NULL(&wr->referent);
+
+	wr->root = NULL;
+}
+
+void zend_weakrefs_notify(zend_object *object) {
+	zend_array *refs = zend_weakrefs_fetch(object, 0);
+
+	if (!refs) {
+		return;
+	}
+
+	zend_hash_index_del(&EG(weakrefs), (zend_ulong) object);
+}
+
+void zend_weakrefs_shutdown() {
+	zend_hash_destroy(&EG(weakrefs));
+}
+
+static zend_object* zend_weakref_new(zend_class_entry *ce) {
+	zend_weakref *wr =
+		(zend_weakref*) 
+			zend_object_alloc(sizeof(zend_weakref), ce);
+
+	zend_object_std_init(&wr->std, ce);
+
+	wr->std.handlers = &zend_weakref_handlers;
+
+	ZVAL_NULL(&wr->referent);
+
+	return &wr->std;
+}
+
+static zend_object* zend_weakref_clone(zval *zv) {
+	zval rv;
+	zend_weakref *wr = zend_weakref_fetch(zv);
+
+	object_init_ex(&rv, Z_OBJCE_P(zv));
+
+	zend_weakref_construct(&rv, &wr->referent);
+
+	return Z_OBJ(rv);
+}
+
+static void zend_weakref_free(zend_object *zo) {
+	zend_weakref *wr = zend_weakref_from(zo);
+
+	if (!wr->root) {
+		zend_object_std_dtor(&wr->std);
+		return;
+	}
+
+	zend_hash_index_del(wr->root, (zend_ulong) wr);
+	zend_object_std_dtor(&wr->std);
```

Comment:
I'd write this as

```c
if (wr->root) {
    zend_hash_index_del(wr->root, (zend_ulong) wr);
}
zend_object_std_dtor(&wr->std);
```

to avoid repeating the common part.